src/Nethermind/Nethermind.Evm/CodeAnalysis/CodeDataAnalyzer.cs

```diff
@@ -0,0 +1,182 @@
+﻿//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Threading;
+
+namespace Nethermind.Evm.CodeAnalysis
+{
+    public class CodeDataAnalyzer : ICodeInfoAnalyzer
+    {
+        private byte[]? _codeBitmap;
+        public byte[] MachineCode { get; set; }
+        
+        public CodeDataAnalyzer(byte[] code)
+        {
+            MachineCode = code;
+        }
+        
+        public bool ValidateJump(int destination, bool isSubroutine)
+        {
+            _codeBitmap ??= CodeDataAnalyzerHelper.CreateCodeBitmap(MachineCode);
+     
+            if (destination < 0 || destination >= MachineCode.Length)
+            {
+                return false;
+            }
+
+            if (!CodeDataAnalyzerHelper.IsCodeSegment(_codeBitmap, destination))
+            {
+                return false;
+            }
+            
+            if (isSubroutine)
+            {
+                return MachineCode[destination] == 0x5c;
+            }
+
+            return MachineCode[destination] == 0x5b;
+        }
+    }
+    
+    public static class CodeDataAnalyzerHelper
+    {
+        private const UInt16 Set2BitsMask = 0b1100_0000_0000_0000;
+        private const UInt16 Set3BitsMask = 0b1110_0000_0000_0000;
+        private const UInt16 Set4BitsMask = 0b1111_0000_0000_0000;
+        private const UInt16 Set5BitsMask = 0b1111_1000_0000_0000;
+        private const UInt16 Set6BitsMask = 0b1111_1100_0000_0000;
+        private const UInt16 Set7BitsMask = 0b1111_1110_0000_0000;
+
+        private static readonly byte[] _lookup = new byte[8] { 0x80, 0x40, 0x20, 0x10, 0x8, 0x4, 0x2, 0x1, };
+        
+        /// <summary>
+        /// Collects data locations in code.
+        /// An unset bit means the byte is an opcode, a set bit means it's data.
+        /// </summary>
+        public static byte[] CreateCodeBitmap(byte[] code)
+        {
+            // The bitmap is 4 bytes longer than necessary, in case the code
+            // ends with a PUSH32, the algorithm will push zeroes onto the
+            // bitvector outside the bounds of the actual code.
+            byte[] bitvec = new byte[(code.Length / 8) + 1 + 4];
```

Comment:
ArrayPool.Rent

Reply:
Do we actually benefit from using it in this scenario? We allocate the array but we pass it to CodeInfo and it leaves with the CodeInfo object. It's the same as the bit array used in the old implementation.

Reply:
Yes, it would have to be disposable, I wonder if that makes sense, you can leave a TODO to analyze it later.